README.md
# Scheduler

![Screenshot](./screenshot.png)

A client-side web app that tracks time spent on specific tasks and organizes them with a tagging system. The app runs entirely in the user's browser; it has no server, and your private information is never transmitted over the internet. Task information is auto-saved to the browser's [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Storage/LocalStorage).

You can use it online, hosted on Github Pages [here](christopherfujino.github.io/scheduler).

Built with:

1. [AngularJS v1.4](https://angularjs.org)
2. [jQuery v2.1.3](http://jquery.com)
3. [Twitter Bootstrap v3.3.4](http://getbootstrap.com)
4. [Thomas Park](http://thomaspark.co/)'s [Lumen Bootstrap theme](http://bootswatch.com/lumen/)
5. [Font Awesome v4.3.0](http://fontawesome.io)


changelog.md
# Changelog

## todo:
* add: support for chrome app

## finished:

version 0.11
* fixed: bug where cumulative tag times wouldn't be refreshed when adding a tag that already exists
* fixed: bug where tagRefresh() only showed time of last save
* added: button to refresh cumulative tag times
* changed: moved tagRefresh() method onto $scope
* fixed: bug where after 24 hours, timestring would reset

version 0.10.1
* fixed: bug where tags wouldn't be removed if task was deleted with tags still attached.
* added: tag times now instantly refresh if a task's elapsed time is adjusted

version 0.10
* added: buttons to add/subtract time from elapsed time

version 0.9
* added: tag system
* added: readout for tags
* changed: tags to object literals, with property for total lapsed time
* changed: streamlined UI interface

version 0.8.2
* added: font-awesome css, bootstrap & jquery js dependencies
* changed: navbar UI streamlined

version 0.8.1
* fixed: app will load saved data from
To squelch this message and adopt the new behavior now, use:

  older versions > 0.6

version 0.8
* have new tasks placed at beginning of arrays
* added: $scope.startToDoTask() to move tasks from taskStore.toDoTasks[] to taskStore.activeTasks[]

version 0.7
* add: bootstrap columns for active and 'to do' tasks
* refactor taskStore as an object, with separate arrays for activeTasks & toDoTasks
* separate view for 'to do' tasks

version 0.6
* add: active tasks now auto-save every x (currently 20) seconds
* change: no longer discard saved tasks with cumulativeTime==0 as these might be 'To Do' tasks

version 0.5
* change: there are now separate buttons: 1) 'Start New Task'; & 2) 'To Do Later'
* fix: discarding of saved tasks with cumulativeTime==0 didn't work because constructors must return objects, now fixed
* migrate to bootstrap.css

version 0.4.1
* caveat: if a task has never been stopped, it will be discarded

version 0.4
* copy taskStore to local storage
* app will auto-save the entire task store to local storage on any stop or delete operation

version 0.3
* add ability to pause and resume task
* add ability to edit task name

version 0.2
* Instead of assigning task as an object literal, version 0.2 will define a task object constructor, instances of which will then be .push()'d into an array $scope.taskStore.

version 0.1
* first working version


scripts/app.js
var isChromeApp = (function(){  // this object will be used to check if we're in a Chrome app or regular web app
  'use strict';
  var theBool = (chrome && chrome.app && chrome.app.runtime) ? true : false;

  return {
    get : function() {
      return theBool;
    }
  };
})();

var app = angular.module('scheduler',[]); // requires angularJS version >= 1.4

app.controller('main', ['$scope', '$timeout', function($scope, $timeout) {
  'use strict';

  $scope.version = '0.11';
  $scope.nameBool = false; // name is clicked?

  $scope.taskStore = {
    version : $scope.version,
    activeTasks : [],
    toDoTasks : [],
    tags : [] // of the form { 'tag' : 'my tag name', 'time' : 5500 }
  };

  var tagRefreshHandler = null;

  $scope.tagRefresh = function() {
    'use strict';
    $timeout.cancel(tagRefreshHandler); // cancel any other scheduled refresh
    tagRefreshHandler = $timeout($scope.tagRefresh, 60000);  // schedule another refresh

    for(var k=0; k<$scope.taskStore.tags.length; k++) { // iterate through each tag of taskStore.tags[]
      var query = $scope.taskStore.tags[k];

      // for queryTag, sum up all cumulativeTime of tasks with this tag
      var ms = 0,
        active = $scope.taskStore.activeTasks; // a simple reference for brevity
      for(var i=0; i<active.length; i++) { // iterate through all active tasks
        for(var j=0; j<active[i].tags.length; j++) { // iterate through all tags of each task
          if(active[i].tags[j] === query.tag) {
            ms += active[i].cumulativeTime + (active[i].initTime ? (new Date().getTime() - active[i].initTime.getTime()) : 0); // if active[i].initTime exists, calculate time lapsed since;
            break;
          }
        }
      }
      query.time = $scope.timeString(ms);
    }
    console.log('Tags refreshed!');
  };

  $scope.timeString = function(x) {
    /*
      this takes a quantity of milliseconds x, and converts it to a string of the
      form Hs:Ms:Ss
    */
    x = Math.floor(x/1000);
    var seconds = x % 60;
    if(seconds < 10) seconds = '0' + seconds;
    x = (x - seconds)/60;

    var minutes = x % 60;
    if (minutes < 10) minutes = '0' + minutes;
    x = (x - minutes)/60;

    var hours = x;  // since no days, don't do a modulo
    if(hours < 10) hours = '0' + hours;

    return hours + ':' + minutes + ':' + seconds;
  };

  if (isChromeApp.get()) { // if we are in a chrome app, use chrome.storage.local
    console.log('reading from chrome.storage.local!');
    chrome.storage.local.get('taskStore', function(getObj) {
      var obj = getObj.taskStore;
      if(obj.version) { // if this is a valid taskStore
        if (obj.version !== $scope.version) { // saved taskStore from an earlier version
          console.log(obj);
          console.log('Warning! Saved data from an earlier version!');
          console.log('Current app version is: ' + $scope.version);
          console.log('Saved data version is: ' + obj.version);
    //      throw new Error("Something went badly wrong!");   // only throw error if saved data API broken
        }
        // if saved data from same version, process
        obj.activeTasks.forEach(function(x) {
          var temp = new Task(x); // temp created to check if Task() returns error object
          if(temp.error !== true) $scope.taskStore.activeTasks.push(temp); // don't push if Task() returns error object
        });
        obj.toDoTasks.forEach(function(x) {
          var temp = new Task(x); // temp created to check if Task() returns error object
          if(temp.error !== true) $scope.taskStore.toDoTasks.push(temp); // don't push if Task() returns error object
        });
        console.log('obj.tags == ');
        console.log(obj.tags);
        if (typeof obj.tags === 'object') { // if tags[] exists from memory, copy it into $scope.taskStore.tags[]
          if (typeof obj.tags[0] === 'string') { // api <= 0.9a
            obj.tags.forEach(function(tag){
              $scope.taskStore.tags.push({
                'tag' : tag,
                'time' : null
              });
            });
          }
          else if (typeof obj.tags[0] === 'object') {
            $scope.taskStore.tags = obj.tags.slice(0);
          }
        }
      }
    } );
  }
  else {  // if not in a chrome app, use localStorage
    console.log('reading from localStorage');
    var temp = localStorage.getItem('taskStore'); // read taskStore from localStorage
    if(temp) { // load saved taskStore if it exists
      var obj = JSON.parse(temp);

      if (!obj.version) { // from version < 0.7, where taskStore was a literal array of Tasks
        obj.forEach(function(x) {
          var temp = new Task(x); // temp created to check if Task() returns error object
          if(temp.error !== true) $scope.taskStore.activeTasks.push(temp); // don't push if Task() returns error object
        });

        console.log('Migrating taskStore from version <= 0.6');

  //      throw new Error("Migrating taskStore from version 0.6");
      }
      else if (obj.version !== $scope.version) { // saved taskStore from an earlier version
        console.log(obj);
        console.log('Warning! Saved data from an earlier version!');
        console.log('Current app version is: ' + $scope.version);
        console.log('Saved data version is: ' + obj.version);
  //      throw new Error("Something went badly wrong!");   // only throw error if saved data API broken
      }
      // if saved data from same version, process
      obj.activeTasks.forEach(function(x) {
        var temp = new Task(x); // temp created to check if Task() returns error object
        if(temp.error !== true) $scope.taskStore.activeTasks.push(temp); // don't push if Task() returns error object
      });
      obj.toDoTasks.forEach(function(x) {
        var temp = new Task(x); // temp created to check if Task() returns error object
        if(temp.error !== true) $scope.taskStore.toDoTasks.push(temp); // don't push if Task() returns error object
      });
      if (obj.tags) { // if tags[] exists from memory, copy it into $scope.taskStore.tags[]
        if (typeof obj.tags[0] === 'string') { // api <= 0.9a
          obj.tags.forEach(function(tag){
            $scope.taskStore.tags.push({
              'tag' : tag,
              'time' : null
            });
          });
        }
        else if (typeof obj.tags[0] === 'object') {
          $scope.taskStore.tags = obj.tags.slice(0);
        }
      }
    }
  }

  $scope.tagRefresh();  // refresh tag times

  $scope.updateStorage = function() {
    if (isChromeApp.get()) { // if we are in a chrome app, use chrome.storage.local
      chrome.storage.local.set( { 'taskStore' : $scope.taskStore } );
      console.log('chrome.storage.local updated!');
      console.log($scope.taskStore);
    }
    else {  // if not in a chrome app, use localStorage
      localStorage.setItem('taskStore', JSON.stringify($scope.taskStore));
      console.log('local storage updated!');
    }
  };

  $scope.newTask = function(name, startNow) {
    var temp = new Task(name, startNow);
    if(temp.error !== true) {
      if(startNow === true) {
        this.taskStore.activeTasks.unshift(temp);
      }
      else {
        this.taskStore.toDoTasks.push(temp); // keep oldest to-do's at top, so push
      }
      this.updateStorage();
    }
  };

  $scope.addTag = function(tag, task) {
    if(task.addTag(tag)) {
      var dupe = false; // is this tag already listed in taskStore.tags[]?
      $scope.taskStore.tags.forEach(function(x){
        if(x.tag===tag) dupe=true;
      });
      if(!dupe) {
        $scope.taskStore.tags.push({'tag': tag, 'time': null}); // only add this tag if it's unique
      }
      $scope.tagRefresh();
      $scope.updateStorage();
    }
  };

  $scope.removeTag = function(tag, task) {
    var i, j;
    if(task.removeTag(tag)){
      var unique = true, // assume this tag was unique, unless another is found
//        task = {},
        loopDone = false; // should we break the outer loop?
      // check if tags should be removed, implement
      for(i=0; !loopDone && i<$scope.taskStore.activeTasks.length; i++) {
        task = $scope.taskStore.activeTasks[i];
        for(j=0; j<task.tags.length; j++) {
          if (task.tags[j] === tag) {
            unique = false;
            loopDone = true; // done looping, tag not unique
            break;
          }
        }
      }

      for(i=0; !loopDone && i<$scope.taskStore.toDoTasks.length; i++) {
        task = $scope.taskStore.toDoTasks[i];
        for(j=0; j<task.tags.length; j++) {
          if (task.tags[j] === tag) {
            unique = false;
            loopDone = true; // done looping, tag not unique
            break;
          }
        }
      }
      console.log('unique = ' + unique);
      if (unique) { // tag was unique, remove it from $scope.taskStore.tags[]
        for(i=0; i<$scope.taskStore.tags.length; i++) {
          if(tag === $scope.taskStore.tags[i].tag) {
            $scope.taskStore.tags.splice(i, 1);
//            console.log('tag ' + tag + ' deleted!');
          }
        }
      }
//      else { console.log('not so unique.'); }
      $scope.updateStorage();
    }
  };

  $scope.startToDoTask = function(index) {
    var activatedTask = $scope.taskStore.toDoTasks.splice(index, 1)[0];
    $scope.taskStore.activeTasks.unshift(activatedTask);
    activatedTask.pause();
  };

  function Task(arg, startNow) { // 1st arg is string for new obj, obj for loaded obj; 2nd arg is bool
    this.tags = [];
    if(typeof arg === 'string') { // create new task with arg being name
      this.name = arg; // arg coming from text input box
      this.cumulativeTime = 0; // time lapsed prior to a pause, integer
      this.initTime = new Date();
      this.date = this.initTime.toDateString();
      this.lapsedTimeString = '';
      if(startNow === true) {
        this.isPaused = false;
        this.timeoutHandler = $timeout(increment, 1000, true, this); // 3rd arg is def, 4th is arg to pass to increment
      }
      else { this.isPaused = true; }
    }
    else if (typeof arg === 'object') { // recreating saved task from local storage
//      if(arg.cumulativeTime === 0) return {error : true};
      this.name = arg.name;
      this.cumulativeTime = arg.cumulativeTime;
      this.date = arg.date;
      this.isPaused = true;
      this.initTime = null;
      this.tags = arg.tags.slice(0);
    }
    this.nameBool = false;  // don't start in name edit mode
    this.toggleName = function() { if (this.nameBool) this.nameBool = false; else this.nameBool = true; };

    function increment(task) {
      task.timeoutHandler = $timeout(increment, 1000, true, task);  // immediately restart timer
      var now = new Date();
      task.setLapsedTime();
      if((now - task.initTime) > 60000) { // save every 60 seconds
        task.save();
      }
    }

    this.adjustTime = function(x) {
      this.cumulativeTime += +x;
      if (this.cumulativeTime < 0) {
        this.cumulativeTime = 0;
      }
      this.setLapsedTime();
      $scope.tagRefresh();
    };

    this.setLapsedTime = function() {
      var temp = this.cumulativeTime + (this.initTime ? (new Date().getTime() - this.initTime.getTime()) : 0); // if this.initTime exists, calculate time lapsed since
      this.lapsedTimeString = $scope.timeString(temp);
    }; this.setLapsedTime(); // and now immediately call this function to initialize lapsedTimeString

    this.pause = function() {
      if(this.isPaused) {
        this.initTime = new Date();
        this.timeoutHandler = $timeout(increment, 1000, true, this); // 3rd arg is def, 4th is arg to pass to increment
        this.isPaused = false;
      }
      else {  // task still running, let's pause
        $timeout.cancel(this.timeoutHandler);
        this.isPaused = true;
        this.save();
        this.initTime = undefined;
      }
    };
    this.remove = function(type, index) {
      if (type !== 'active' && type !== 'toDo') {
        throw new Error('unknown type "' + type + '" in Task.remove()');
      } else {
        var collection = (type === 'active') ? $scope.taskStore.activeTasks : $scope.taskStore.toDoTasks;
        while(this.tags.length > 0) { // length of this.tags[] will change as tags are removed
          $scope.removeTag(this.tags[0], this); // call $scope.removeTag() so that taskStore's tags will be cleaned
        }
        collection.splice(index, 1);
        $scope.updateStorage();
      }
    };
    this.addTag = function(tag) {
      var isRepeat = false;
      this.tags.forEach(function(x) {
        if (x===tag) isRepeat = true;
      });
      if(!isRepeat && tag !== '') {
        this.tags.push(tag);
        return true;
      }
      else return false;
    };
    this.removeTag = function(tag) {
      var tagToRemove = null;
      this.tags.forEach(function(x, index) {
        if (x===tag) tagToRemove = index;
      });
      if(tagToRemove !== null) {
        this.tags.splice(tagToRemove, 1);
        return true;
      }
      else {
        return false;
      }
    };
    this.save = function() {
      var now = new Date();
      this.cumulativeTime += now.getTime() - this.initTime.getTime();
      this.initTime = now;
      $scope.updateStorage();
    };
  }

}]);
